### PR TITLE
User contacts improvement

### PR DESCRIPTION
**Pull request by ice-myles**
Contacts table processor was added to handle contact referrals from contacts-table topic. Contact notification was changed accordingly to send notifications about new contact to user.

Related to [eskimo PR#6](https://github.com/ice-blockchain/eskimo/pull/6)

### CODE REVIEW

File: notifications/contract.go [hunk 1]
```diff
@@ -226,6 +226,9 @@ type (
 	enabledRolesSource struct {
 		*processor
 	}
+	contactsTableSource struct {
```
Comment (ice-ares) [inline notifications/contract.go]: agendaContacts

File: notifications/notification_type_new_contact.go [hunk 2]
```diff
@@ -16,57 +16,51 @@ import (
 	"github.com/ice-blockchain/wintr/time"
 )
 
-func (r *repository) sendNewContactNotification(ctx context.Context, us *users.UserSnapshot) error { //nolint:funlen,gocognit,gocyclo,revive,cyclop // .
+func (r *repository) sendNewContactNotification(ctx context.Context, contact *users.Contact) error { //nolint:funlen,gocognit // .
```
Comment (ice-ares) [inline notifications/notification_type_new_contact.go]: AgendaContact

File: notifications/notification_type_new_contact.go [hunk 3]
```diff
@@ -135,13 +129,13 @@ func (r *repository) getPushNotificationTokensForNewContactNotification(
 								   AND dm.user_id = u.user_id
 								   AND dm.push_notification_token IS NOT NULL 
 								   AND dm.push_notification_token != ''
-						WHERE u.agenda_phone_number_hashes IS NOT NULL
-						  AND POSITION($1 IN u.agenda_phone_number_hashes) != 0
+						WHERE u.user_id = $1
```
Comment (ice-ares) [inline notifications/notification_type_new_contact.go:128]: Do you send  a message for each (userId,contactUserId) pair u find in db, from eskimo?

Meaning contactUserId is the same, but userId is different, cuz there could be multiple users that have that contact in their agenda, not just the user that generated the request in eskimo

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: notifications/notification_type_new_contact.go (hunk 3 above)
Comment (ice-myles) [reply notifications/notification_type_new_contact.go:128]: Why userId should be different?
User has agenda phone number hashes, we send modify request for this user with `phoneNumberHash1,phoneNumberHash2,phoneNumberHash3`, then we should find the users that has these phone number hashes, then send {userID, user1::phoneNumberHash1}, {userID, user2::phoneNumberHash2}, {userID, user3::phoneNumberHash3} pairs only (::phoneNumberHash I use is to show the correspondence between userID and phonehNumberHash).

Comment (ice-ares) [reply notifications/notification_type_new_contact.go:128]: Then u should change it to detect all userId values for that contactUserId, cuz we need to detect all the users that have that other user in their contacts, not just the one requesting it

File: news/DDL.sql [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                    created_at            TIMESTAMP NOT NULL,
+                    updated_at            TIMESTAMP NOT NULL,
+                    notification_channels TEXT,
+                    id                    TEXT NOT NULL,
+                    type                  TEXT NOT NULL,
+                    language              TEXT NOT NULL,
+                    title                 TEXT NOT NULL,
+                    image_url             TEXT NOT NULL,
+                    url                   TEXT NOT NULL UNIQUE,
+                    views                 BIGINT NOT NULL DEFAULT 0,
```
Comment (ice-ares) [inline news/DDL.sql]: Move this after last timestamp

File: news/DDL.sql [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                    PRIMARY KEY(language,id)
+                    );
+CREATE INDEX IF NOT EXISTS most_recent_news_lookup_ix ON news (language, type, created_at);
```
Comment (ice-ares) [inline news/DDL.sql]: Add DESC to created at

File: news/news_create.go [hunk 6]
```diff
@@ -48,23 +48,19 @@ func (r *repository) insertNews(ctx context.Context, news []*TaggedNews) error {
...
-		values = append(values, fmt.Sprintf(`(:created_at%[1]v, :created_at%[1]v, :notification_channels%[1]v, :id%[1]v, :type%[1]v, :language%[1]v, :title%[1]v, :image_url%[1]v, :url%[1]v)`, ix)) //nolint:lll // .
+		args = append(args, nws.CreatedAt.Time, nws.UpdatedAt.Time, nws.NotificationChannels.NotificationChannels, nws.ID, nws.Type, nws.Language,
+			nws.Title, nws.ImageURL, nws.URL,
+		)
+		values = append(values, fmt.Sprintf("($%[1]v,$%[2]v,$%[3]v,$%[4]v,$%[5]v,$%[6]v,$%[7]v,$%[8]v,$%[9]v)",
+			fields*ix+1, fields*ix+2, fields*ix+3, fields*ix+4, fields*ix+5, fields*ix+6, fields*ix+7, fields*ix+8, fields*ix+9)) //nolint:gomnd // .
 	}
 	sql := fmt.Sprintf(`INSERT INTO news (CREATED_AT, UPDATED_AT, NOTIFICATION_CHANNELS, ID, TYPE, LANGUAGE, TITLE, IMAGE_URL, URL) VALUES %v`, strings.Join(values, ",")) //nolint:lll // .
-	if err := storage.CheckSQLDMLErr(r.db.PrepareExecute(sql, params)); err != nil {
-		return errors.Wrapf(detectAndParseDuplicateDatabaseError(err), "failed to insert news %#v", news)
```
Comment (ice-ares) [inline news/news_create.go]: U dont need this?

File: news/news_modify.go [hunk 7]
```diff
@@ -131,32 +137,25 @@ func (r *repository) IncrementViews(ctx context.Context, newsID, language string
...
-	}
-	if err := r.sendNewsViewedMessage(ctx, tuple); err != nil {
-		bErr := errors.Wrapf(err, "failed to sendNewsViewedMessage for %#v", tuple)
-		ops = append(make([]tarantool.Op, 0, 1), tarantool.Op{Op: "-", Field: 9, Arg: uint64(1)}) //nolint:gomnd // That's the `views` column index.
-		if err = storage.CheckNoSQLDMLErr(r.db.UpdateTyped("NEWS", "pk_unnamed_NEWS_2", []any{language, newsID}, ops, &[]*News{})); err != nil {
-			return multierror.Append(bErr, errors.Wrapf(err, "[rollback]failed to decrement news views count for newsId:%v,language:%v", newsID, language)).ErrorOrNil() //nolint:lll,wrapcheck // .
-		}
-		if err = storage.CheckNoSQLDMLErr(r.db.DeleteTyped("NEWS_VIEWED_BY_USERS", "pk_unnamed_NEWS_VIEWED_BY_USERS_1", []any{language, tuple.NewsID, tuple.UserID}, &[]*ViewedNews{})); err != nil { //nolint:lll // .
-			return multierror.Append(bErr, errors.Wrapf(err, "[rollback]failed to delete NEWS_VIEWED_BY_USERS%#v", tuple)).ErrorOrNil() //nolint:wrapcheck // .
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline news/news_modify.go]: Follow the same pattern as before

File: notifications/email_notifications.go [hunk 8]
```diff
@@ -31,23 +29,19 @@ func (r *repository) sendEmailNotification(ctx context.Context, en *emailNotific
...
-		en.en.From.Name = internationalizedEmailDisplayNames["en"]
-	}
-	if err := r.emailClient.Send(ctx, en.en, email.Participant{Name: en.displayName, Email: en.sn.NotificationChannelValue}); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to send email notification:%#v, desired to be sent:%#v", en.en, en.sn),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_NOTIFICATIONS", "pk_unnamed_SENT_NOTIFICATIONS_1", &en.sn.sentNotificationPK, &[]*sentNotification{})), //nolint:lll // .
-				"failed to delete SENT_NOTIFICATIONS as a rollback for %#v", en.sn),
-		).ErrorOrNil()
-	}
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/email_notifications.go]: Follow the same pattern as before

File: notifications/inapp_notifications.go [hunk 9]
```diff
@@ -23,42 +22,34 @@ type (
...
-	}
 
-	if err := r.personalInAppFeed.Send(ctx, in.in, in.sn.UserID); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to send inApp notification:%#v, desired to be sent:%#v", in.in, in.sn),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_NOTIFICATIONS", "pk_unnamed_SENT_NOTIFICATIONS_1", &in.sn.sentNotificationPK, &[]*sentNotification{})), //nolint:lll // .
-				"failed to delete SENT_NOTIFICATIONS as a rollback for %#v", in.sn),
-		).ErrorOrNil()
-	}
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/inapp_notifications.go]: Follow the same pattern as before

File: notifications/inapp_notifications.go [hunk 10]
```diff
@@ -23,42 +22,34 @@ type (
...
-	}
 
-	if err := r.globalInAppFeed.Send(ctx, bin.in, bin.sa.NotificationChannelValue); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to broadcast inApp notification:%#v, desired to be sent:%#v", bin.in, bin.sa),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_ANNOUNCEMENTS", "pk_unnamed_SENT_ANNOUNCEMENTS_1", &bin.sa.sentAnnouncementPK, &[]*sentAnnouncement{})), //nolint:lll // .
-				"failed to delete SENT_ANNOUNCEMENTS as a rollback for %#v", bin.sa),
-		).ErrorOrNil()
-	}
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/inapp_notifications.go]: Follow the same pattern as before

File: notifications/push_notifications.go [hunk 11]
```diff
@@ -147,27 +139,27 @@ func (r *repository) sendPushNotification(ctx context.Context, pn *pushNotificat
...
-		return errors.Wrapf(err, "failed to insert %#v", pn.sn)
-	}
-	responder := make(chan error, 1)
-	defer close(responder)
-	r.pushNotificationsClient.Send(ctx, pn.pn, responder)
-	if err := <-responder; err != nil {
-		var cErr error
-		if errors.Is(err, push.ErrInvalidDeviceToken) {
-			cErr = r.clearInvalidPushNotificationToken(ctx, pn.sn.UserID, pn.pn.Target)
+	err := storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/push_notifications.go]: Follow the same pattern as before

File: notifications/push_notifications.go [hunk 12]
```diff
@@ -210,16 +197,12 @@ func (r *repository) broadcastPushNotification(ctx context.Context, bpn *broadca
...
-	if err := r.pushNotificationsClient.Broadcast(ctx, bpn.pn); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to broadcast push notification:%#v, desired to be sent:%#v", bpn.pn, bpn.sa),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_ANNOUNCEMENTS", "pk_unnamed_SENT_ANNOUNCEMENTS_1", &bpn.sa.sentAnnouncementPK, &[]*sentAnnouncement{})), //nolint:lll // .
-				"failed to delete SENT_ANNOUNCEMENTS as a rollback for %#v", bpn.sa),
-		).ErrorOrNil()
-	}
 
-	return nil
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/push_notifications.go]: Follow the same pattern as before

**Review by ice-ares — CHANGES_REQUESTED** [review]

**Pull request merged**